src/uct/ib/dc/dc_mlx5_ep.c

```diff
@@ -1068,43 +1077,79 @@ uct_dc_mlx5_iface_dci_do_pending_tx(ucs_arbiter_t *arbiter,
 
     if (!uct_dc_mlx5_iface_dci_ep_can_send(ep)) {
         /* Deschedule the group even if FC is the only resource, which
-         * is missing. It will be scheduled again when credits arrive.
-         * We can't desched group with rand policy if non FC resources are
-         * missing, since it's never scheduled again. */
-        if (uct_dc_mlx5_iface_is_dci_rand(iface) &&
-            uct_rc_fc_has_resources(&iface->super.super, &ep->fc)) {
-            return UCS_ARBITER_CB_RESULT_RESCHED_GROUP;
-        } else {
-            return UCS_ARBITER_CB_RESULT_DESCHED_GROUP;
-        }
+         * is missing. It will be scheduled again when credits arrive. */
+        return UCS_ARBITER_CB_RESULT_DESCHED_GROUP;
     }
 
     ucs_assertv(!uct_rc_iface_has_tx_resources(&iface->super.super),
                 "pending callback returned error but send resources are available");
     return UCS_ARBITER_CB_RESULT_STOP;
 }
 
+ucs_arbiter_cb_result_t
+uct_dc_mlx5_iface_dci_do_rand_pending_tx(ucs_arbiter_t *arbiter,
+                                         ucs_arbiter_elem_t *elem,
+                                         void *arg)
+{
+    uct_pending_req_t *req     = ucs_container_of(elem, uct_pending_req_t, priv);
+    uct_dc_mlx5_ep_t *ep       = uct_dc_mlx5_pending_req_priv(req)->ep;
+    uct_dc_mlx5_iface_t *iface = ucs_derived_of(ep->super.super.iface,
+                                                uct_dc_mlx5_iface_t);
+    ucs_status_t status;
+
+    if (!uct_rc_iface_has_tx_resources(&iface->super.super)) {
```

Comment:
looks like uct_dc_mlx5_iface_dci_do_rand_pending_tx() and uct_dc_mlx5_iface_dci_do_dcs_pending_tx() have similar structure, can we extract common code to helper function?